### PR TITLE
Issue 462 - Introduce RepoVariant type to manage metadata

### LINKED ISSUE #462: store metadata as their original type instead of casting it all to string, and guess-cast it back
### Description
At the moment the metadata constructor takes in strings for the values. Then use boost to guess their original types and cast them back.

This is a bit backwards as very often we're [casting the original type to string](https://github.com/3drepo/3drepobouncer/blob/master/bouncer/src/repo/manipulator/modelconvertor/import/repo_model_import_3drepo.cpp#L63) in order to fit the function definition here, only to cast it back to the [original type](https://github.com/3drepo/3drepobouncer/blob/master/bouncer/src/repo/core/model/bson/repo_bson_factory.cpp#L223):

It's probably better if we use a variant type (like https://en.cppreference.com/w/cpp/utility/variant ) to keep the type until the very end.

### Goal
- Create a `RepoVariant` class, inheriting std library variant type
- Update the metadata constructor to take in `RepoVariant for value type instead of strings
  - Remove the lexical cast attempts
- Update all importers with metadata to stop the casting their data to string
  - If there are any importers that are providing the value as string already, attempt the lexical cast there

### PR DESCRIPTION
**Pull request by sbanethia**
This fixes #462 

#### Description
Added RepoVariant class to accept multiple data types in the meta data and removed unwanted casting to std::string and removed lexical cast to std::string

### CODE REVIEW

File: bouncer/src/repo/lib/datastructure/repo_variant.h [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+#include <boost/variant.hpp>
+#include <boost/core/demangle.hpp>
+#include <string>
+#include "../../repo_bouncer_global.h"
+
+namespace repo {
+
+    namespace lib {
+        typedef REPO_API_EXPORT boost::variant<int, double, std::string, bool, boost::blank, uint64_t, float> repoVariant;
+        class REPO_API_EXPORT RepoVariant:public repoVariant {
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/lib/datastructure/repo_variant.h]: Changes done.

File: bouncer/src/repo/lib/datastructure/repo_variant.h [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+#include <iostream>
+#include <boost/variant.hpp>
+#include <boost/core/demangle.hpp>
+#include <string>
+#include "../../repo_bouncer_global.h"
+
+namespace repo {
+
+    namespace lib {
+        typedef REPO_API_EXPORT boost::variant<int, double, std::string, bool, boost::blank, uint64_t, float> repoVariant;
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/lib/datastructure/repo_variant.h]: Changes done.

File: bouncer/src/repo/core/model/bson/repo_bson_factory.cpp [hunk 3]
```diff
@@ -210,33 +185,25 @@ MetadataNode RepoBSONFactory::makeMetaDataNode(
...
+		if (!key.empty() && !value.isEmpty())
 		{
 			RepoBSONBuilder metaEntryBuilder;
 			metaEntryBuilder.append(REPO_NODE_LABEL_META_KEY, key);
 			//Check if it is a number, if it is, store it as a number
 
-			try {
-				long long valueInt = boost::lexical_cast<long long>(value);
-				metaEntryBuilder.append(REPO_NODE_LABEL_META_VALUE, valueInt);
+			if ("int" == value.getVariantType()) {
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/core/model/bson/repo_bson_factory.cpp]: Changes done.

File: bouncer/src/repo/lib/datastructure/repo_variant.h [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+            RepoVariant(std::string& other) : repoVariant(other) {};
+
+
+            RepoVariant& operator=(const std::string& str) {
+                repoVariant::operator=(str);
+                return *this;
+            }
+            
+            std::string getVariantType() const {
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/lib/datastructure/repo_variant.h]: Changes done.

File: bouncer/src/repo/lib/datastructure/repo_variant.h [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+                    // Handle the case where the type is not supported
+                    return "Unsupported type";
+                }
+            }
+
+            bool isEmpty() const {
+                return this->which() == 0;
+            }
+
+            bool isBool() const {
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/lib/datastructure/repo_variant.h]: Changes done.

File: bouncer/src/repo/lib/datastructure/repo_variant.h [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    namespace lib {
+        typedef REPO_API_EXPORT boost::variant<int, double, std::string, bool, boost::blank, uint64_t, float> repoVariant;
+        class REPO_API_EXPORT RepoVariant:public repoVariant {
+        public:
+           using boost::variant<int, double, std::string, bool, boost::blank, uint64_t, float>::variant;  // Inherit constructors
+
+            // Default constructor
+            RepoVariant() = default;
+
+            RepoVariant(const std::string& other) : repoVariant(other) {};
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/lib/datastructure/repo_variant.h:54]: Changes are done as per the suggestion.

File: bouncer/src/repo/lib/datastructure/repo_variant.h [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+
+
+            RepoVariant& operator=(const std::string& str) {
+                repoVariant::operator=(str);
+                return *this;
+            }
+            
+            std::string getVariantType() const {
+                try {
+                    return boost::core::demangle(typeid(*this).name());
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/lib/datastructure/repo_variant.h]: The changes are done as per the suggestion.

File: bouncer/src/repo/manipulator/modelconvertor/import/ifcHelper/repo_ifc_helper_parser.cpp [hunk 8]
```diff
@@ -52,7 +52,8 @@ void convertTreeToNodes(
 		}
 
 		if (tree.meta.size()) {
-			metaSet.insert(new repo::core::model::MetadataNode(repo::core::model::RepoBSONFactory::makeMetaDataNode(tree.meta, tree.name, metaParents)));
+			std::unordered_map<std::string, repo::lib::RepoVariant> metaData = repo::lib::RepoVariant().convertToRepoVariant(tree.meta);
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/manipulator/modelconvertor/import/ifcHelper/repo_ifc_helper_parser.cpp]: Changes are done as per the suggestion.

File: bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/geometry_collector.cpp [hunk 9]
```diff
@@ -383,7 +383,8 @@ repo::core::model::MetadataNode*  GeometryCollector::createMetaNode(
 	const repo::lib::RepoUUID &parentId,
 	const  std::unordered_map<std::string, std::string> &metaValues
 ) {
-	return new repo::core::model::MetadataNode(repo::core::model::RepoBSONFactory::makeMetaDataNode(metaValues, name, { parentId }));
+	std::unordered_map<std::string, repo::lib::RepoVariant> metaData = repo::lib::RepoVariant().convertToRepoVariant(metaValues);
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/geometry_collector.cpp]: Changes are done as per the suggestion.

File: bouncer/src/repo/manipulator/modelconvertor/import/repo_model_import_assimp.cpp [hunk 10]
```diff
@@ -610,8 +610,10 @@ repo::core::model::MetadataNode* AssimpModelImport::createMetadataRepoNode(
 			}
 		}
 
+
+		std::unordered_map<std::string, repo::lib::RepoVariant> metaData = repo::lib::RepoVariant().convertToRepoVariant(metaEntries);
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/manipulator/modelconvertor/import/repo_model_import_assimp.cpp]: Changes are done as per the suggestion.

File: bouncer/src/repo/lib/datastructure/repo_variant.h [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+            std::string toString() const {
+                try {
+                    return boost::get<std::string>(*this);
+                }
+                catch (const boost::bad_get& e) {
+                    throw std::runtime_error("Failed to convert variant to std::string");
+                }
+            }
+
+            static std::unordered_map<std::string, RepoVariant> convertToRepoVariant(const std::unordered_map<std::string, std::string>& inputMap) {
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/lib/datastructure/repo_variant.h:127]: Changes are done as per the suggestion.

**Review by carmenfan — CHANGES_REQUESTED** [review]
@sbanethia I'm throwing this back to address feedback, re-request for a review when it's ready

File: ifcUtils/repo_ifc_utils_tree_parser.cpp [hunk 12]
```diff
@@ -1268,7 +1269,7 @@ void  repo::ifcUtility::SCHEMA_NS::TreeParser::generateClassificationInformation
 				if (refSource->hasSource()) {
 #endif
 					classificationName = refSource->Name();
-					metaValues[constructMetadataLabel("Name", classificationName)] = classificationName;
+					metaValues[constructMetadataLabel("Name", classificationName)] = classificationName.c_str();
```
Comment (carmenfan) [inline ifcUtils/repo_ifc_utils_tree_parser.cpp:1272]: same here

File: test/src/system/st_3drepobouncerClient.cpp [hunk 13]
```diff
@@ -104,11 +104,12 @@ static std::string produceUploadArgs(
 	const std::string &filePath,
 	const std::string &configPath = getConnConfig())
 {
-	return  getClientExePath()
+	std::string path = getClientExePath()
```
Comment (carmenfan) [inline test/src/system/st_3drepobouncerClient.cpp:107]: why do you need this change?

File: test/src/unit/repo/core/model/bson/ut_repo_bson_factory.cpp [hunk 14]
```diff
@@ -284,31 +284,34 @@ TEST(RepoBSONFactoryTest, MakeMaterialNodeTest)
...
 {
-	std::vector<std::string> keys({ "one", "two", "three", "four", "five" }), values({ "!", "!!", "!!!", "!!!!", "!!!!!" });
-
 	std::string name = "MetaTest";
-
-	MetadataNode metaNode = RepoBSONFactory::makeMetaDataNode(keys, values, name);
-
+	std::unordered_map<std::string, repo::lib::RepoVariant> metaDataUnMap;
+
+	metaDataUnMap["one"]=std::string("!");
```
Comment (carmenfan) [inline test/src/unit/repo/core/model/bson/ut_repo_bson_factory.cpp:290]: can we test some non string values please.

File: test/src/unit/repo/core/model/bson/ut_repo_node_metadata.cpp [hunk 15]
```diff
@@ -30,13 +30,12 @@ using namespace repo::core::model;
 static MetadataNode makeRandomMetaNode()
 {
 	uint32_t nItems = rand() % 10 + 1;
-	std::vector<std::string> keys, values;
+	std::unordered_map<std::string,repo::lib::RepoVariant> metaDataUnMap;
```
Comment (carmenfan) [inline test/src/unit/repo/core/model/bson/ut_repo_node_metadata.cpp:33]: same here, now we support a variant, the data should not be just strings.

File: test/src/unit/repo_test_utils.h [hunk 16]
```diff
@@ -223,6 +223,7 @@ static bool projectHasGeometryWithMetadata(std::string dbName, std::string proje
 			for (auto m : meshes) {
 				auto meshNode = dynamic_cast<repo::core::model::MeshNode*>(m);
 				sharedIdToMeshNode[meshNode->getSharedID()] = meshNode;
+				repoInfo << "meshNode SharedId=" << meshNode->getSharedID().toString();
```
Comment (carmenfan) [inline test/src/unit/repo_test_utils.h:226]: remove debug logging

File: test/src/unit/repo_test_utils.h [hunk 17]
```diff
@@ -237,13 +238,11 @@ static bool projectHasGeometryWithMetadata(std::string dbName, std::string proje
 					// toString will stringify the underlying type of the value, in the same
 					// way it is stringified for the frontend.
 
-					auto aValue = entry.toMongoElement().Obj().getField(REPO_NODE_LABEL_META_VALUE).toString(false);
-
+					auto aValue = entry.toMongoElement().Obj().getField(REPO_NODE_LABEL_META_VALUE).String();
```
Comment (carmenfan) [inline test/src/unit/repo_test_utils.h:241]: not sure why this is changed

We haven't really changed the way we store the field in mongo?

**Review by carmenfan — CHANGES_REQUESTED** [review]

File: bouncer/src/repo/core/model/bson/repo_bson_builder.h [hunk 18]
```diff
@@ -63,16 +64,45 @@ namespace repo {
...
+					case repo::lib::RepoDataType::BOOL:
+						mongo::BSONObjBuilder::append(label, repoVariant.toBool());
+						break;
+					case repo::lib::RepoDataType::DOUBLE:
+						mongo::BSONObjBuilder::append(label, repoVariant.toDouble());
+						break;
+					case repo::lib::RepoDataType::INT:
+						mongo::BSONObjBuilder::append(label, repoVariant.toInt());
+						break;
+					case repo::lib::RepoDataType::OTHER:
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/core/model/bson/repo_bson_builder.h:95]: Have removed the OTHER datatype and have implemented the warning message in the default case.

File: bouncer/src/repo/core/model/bson/repo_bson_factory.cpp [hunk 19]
```diff
@@ -210,40 +185,19 @@ MetadataNode RepoBSONFactory::makeMetaDataNode(
...
 		std::string key = sanitiseKey(entry.first);
-		std::string value = entry.second;
-
-		if (!key.empty() && !value.empty())
+		repo::lib::RepoVariant value = entry.second;
+		if (!key.empty() && !value.isEmpty())
 		{
 			RepoBSONBuilder metaEntryBuilder;
 			metaEntryBuilder.append(REPO_NODE_LABEL_META_KEY, key);
 			//Check if it is a number, if it is, store it as a number
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/core/model/bson/repo_bson_factory.cpp:195]: Removed the comment as per the suggestion.

File: bouncer/src/repo/core/model/bson/repo_bson_builder.h [hunk 20]
```diff
@@ -35,6 +35,7 @@
 #include "../../../lib/datastructure/repo_matrix.h"
 #include "../../../lib/datastructure/repo_uuid.h"
 #include "repo_bson.h"
+#include <repo/lib/datastructure/repo_variant.h>
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/core/model/bson/repo_bson_builder.h:38]: We have made the changes as per your suggestion and will follow this in future also.

File: bouncer/src/repo/lib/datastructure/repo_variant.h [hunk 21]
```diff
@@ -0,0 +1,108 @@
...
+#include <iostream>
+#include <boost/variant.hpp>
+#include <string>
+#include "../../repo_bouncer_global.h"
+#include "repo/lib/repo_log.h"
+
+namespace repo {
+
+    namespace lib {
+        using repoVariant = boost::variant<int, double, std::string, bool, uint64_t, float, long>;
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/lib/datastructure/repo_variant.h:11]: Changes done as per the suggession.

File: test/src/unit/repo_test_utils.h (hunk 16 above)
Comment (sbanethia) [reply test/src/unit/repo_test_utils.h:226]: Debugging logs are removed.

File: test/src/system/st_3drepobouncerClient.cpp (hunk 13 above)
Comment (sbanethia) [reply test/src/system/st_3drepobouncerClient.cpp:107]: This change was done for debugging purposes. Removing this unnecessary path variable.

File: bouncer/src/repo/lib/datastructure/repo_variant.h [hunk 22]
```diff
@@ -0,0 +1,108 @@
...
+            RepoVariant(const uint64_t& data) : repoVariant(data){};
+
+            RepoVariant(const float& data) : repoVariant(data){};
+
+            RepoVariant(const std::string& data) : repoVariant(data){};
+
+            RepoVariant(const long& data) : repoVariant(data){};
+
+            repo::lib::RepoDataType getVariantType() {
+                switch (which()) {
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/lib/datastructure/repo_variant.h:34]: Changes are done as per the suggestion.

File: bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_dgn.cpp [hunk 23]
```diff
@@ -137,7 +137,7 @@ bool DataProcessorDgn::doDraw(OdUInt32 i, const OdGiDrawable* pDrawable)
 	if (!collector->hasMeta(groupID)) {
 		auto meta = extractXMLLinkages(previousItem);
 		if (!layerName.empty()) {
-			meta["Layer Name"] = layerName;
+			meta["Layer Name"] = layerName.c_str();
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_dgn.cpp:140]: This was not necessary and hence removed.

File: bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_nwd.cpp [hunk 24]
```diff
@@ -137,31 +137,31 @@ void convertColor(OdString color, std::vector<float>& dest)
...
 {
 	setMetadataValue(category, key, convertToStdString(value), metadata);
 }
 
 // This specialisation is because v140 doesn't support constexpr and will attempt to find a to_string overload regardless of the if-statement
 template <>
-void setMetadataValue(const OdString& category, const OdString& key, const double& value, std::unordered_map<std::string, std::string>& metadata)
+void setMetadataValue(const OdString& category, const OdString& key, const double& value, std::unordered_map<std::string, repo::lib::RepoVariant>& metadata)
 {
 	setMetadataValue(category, key, std::to_string(value), metadata);
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_nwd.cpp:158]: Removed to_string as this is of no use now.

File: bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_rvt.cpp [hunk 25]
```diff
@@ -370,7 +370,7 @@ void DataProcessorRvt::fillMeshData(const OdGiDrawable* pDrawable)
 
 void DataProcessorRvt::fillMetadataById(
 	OdBmObjectId id,
-	std::unordered_map<std::string, std::string>& metadata)
+	std::unordered_map<std::string, repo::lib::RepoVariant>& metadata)
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_rvt.cpp:373]: Removed unwanted string casting.

File: bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/geometry_collector.cpp [hunk 26]
```diff
@@ -348,7 +348,11 @@ repo::core::model::RepoNodeSet GeometryCollector::getMeshNodes(const repo::core:
 						auto itPtr = elementToMetaNode.find(meshGroupEntry.first);
 						auto metaParent = partialObject ? parentId : meshNode.getSharedID();
 						if (itPtr == elementToMetaNode.end()) {
-							auto metaNode = createMetaNode(meshGroupEntry.first, {}, idToMeta[meshGroupEntry.first]);
+							std::unordered_map<std::string, repo::lib::RepoVariant> idToRepoVariantMeta;
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/geometry_collector.cpp:351]: You were right the copy was not needed hence removing it.

File: bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/geometry_collector.cpp [hunk 27]
```diff
@@ -394,7 +398,11 @@ repo::core::model::TransformationNode*  GeometryCollector::createTransNode(
 	auto transNode = new repo::core::model::TransformationNode(repo::core::model::RepoBSONFactory::makeTransformationNode(repo::lib::RepoMatrix(), name, { parentId }));
 	if (idToMeta.find(id) != idToMeta.end()) {
 		if (elementToMetaNode.find(id) == elementToMetaNode.end()) {
-			auto metaNode = createMetaNode(name, transNode->getSharedID(), idToMeta[id]);
+			std::unordered_map<std::string, repo::lib::RepoVariant> idToRepoVariantMeta;
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/geometry_collector.cpp:401]: Removing unwanted copy of idToMeta to idToRepoVariantMeta as it is not needed.

File: bouncer/src/repo/manipulator/modelconvertor/import/repo_model_import_assimp.cpp [hunk 28]
```diff
@@ -572,26 +572,27 @@ repo::core::model::MetadataNode* AssimpModelImport::createMetadataRepoNode(
 			switch (currentValue.mType)
 			{
 			case AI_BOOL:
-				metaEntries[key] = std::to_string(*(static_cast<bool *>(currentValue.mData)));
+				metaEntries[key] = currentValue.mData;
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/manipulator/modelconvertor/import/repo_model_import_assimp.cpp:575]: Removing only string casting as it is not needed.

File: bouncer/src/repo/manipulator/modelconvertor/import/repo_model_import_assimp.cpp [hunk 29]
```diff
@@ -572,26 +572,27 @@ repo::core::model::MetadataNode* AssimpModelImport::createMetadataRepoNode(
...
+				metaEntries[key]=currentValue.mData;
 				break;
 
 			case AI_AISTRING:
+				//Need to change this to string and then use the repovariant
 				val = (static_cast<aiString *>(currentValue.mData))->C_Str();
 
 				if (val.compare(key)) {
-					metaEntries[key] = val;
+					metaEntries[key] = val.c_str();
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/manipulator/modelconvertor/import/repo_model_import_assimp.cpp:595]: c_str() is not needed it was done for some testing purpose but should be removed, hence removing it now.

File: ifcUtils/repo_ifc_utils_tree_parser.cpp [hunk 30]
```diff
@@ -293,7 +294,7 @@ repo::ifcUtility::SCHEMA_NS::TreeParser::Actions_t repo::ifcUtility::SCHEMA_NS::
 			units = processUnits(propVal->Unit()).second;
 		}
 
-		metaValues[constructMetadataLabel(propVal->Name(), metaPrefix, units)] = value;
+		metaValues[constructMetadataLabel(propVal->Name(), metaPrefix, units)] = value.c_str();
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply ifcUtils/repo_ifc_utils_tree_parser.cpp:297]: c_str() is not needed it was done for some testing purpose but should be removed, hence removing it now.

File: ifcUtils/repo_ifc_utils_tree_parser.cpp (hunk 12 above)
Comment (sbanethia) [reply ifcUtils/repo_ifc_utils_tree_parser.cpp:1272]: c_str() is not needed it was done for some testing purpose but should be removed, hence removing it now.

File: ifcUtils/repo_ifc_utils_globals.h [hunk 31]
```diff
@@ -20,6 +20,7 @@
 #include <unordered_map>
 #include <cstdint>
 #include <string>
+#include <repo/lib/datastructure/repo_variant.h>
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply ifcUtils/repo_ifc_utils_globals.h:23]: This is resolved as suggested.

File: test/src/unit/repo_test_utils.h (hunk 17 above)
Comment (sbanethia) [reply test/src/unit/repo_test_utils.h]: This change was not necessary this was done for checking purpose hence reverting this changes now.

File: test/src/unit/repo/core/model/bson/ut_repo_bson_factory.cpp (hunk 14 above)
Comment (sbanethia) [reply test/src/unit/repo/core/model/bson/ut_repo_bson_factory.cpp]: Now we have changed this to integer and are testing this.

File: test/src/unit/repo/core/model/bson/ut_repo_node_metadata.cpp (hunk 15 above)
Comment (sbanethia) [reply test/src/unit/repo/core/model/bson/ut_repo_node_metadata.cpp:33]: Now we have changed this to integer and are testing this.

File: bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_nwd.cpp [hunk 32]
```diff
@@ -137,31 +137,31 @@ void convertColor(OdString color, std::vector<float>& dest)
 }
 
 template <class T>
-void setMetadataValue(const OdString& category, const OdString& key, const T& value, std::unordered_map<std::string, std::string>& metadata)
+void setMetadataValue(const OdString& category, const OdString& key, const T& value, std::unordered_map<std::string, repo::lib::RepoVariant>& metadata)
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_nwd.cpp:140]: Changes have been done as per the suggession.

File: bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_rvt.cpp [hunk 33]
```diff
@@ -482,11 +482,10 @@ void DataProcessorRvt::fillMetadataByElemPtr(
...
 }
 
-std::unordered_map<std::string, std::string> DataProcessorRvt::fillMetadata(OdBmElementPtr element)
+std::unordered_map<std::string, repo::lib::RepoVariant> DataProcessorRvt::fillMetadata(OdBmElementPtr element)
 {
-	std::unordered_map<std::string, std::string> metadata;
-	metadata[REVIT_ELEMENT_ID] = std::to_string((OdUInt64)element->objectId().getHandle());
-
+	std::unordered_map<std::string, repo::lib::RepoVariant> metadata;
+	metadata[REVIT_ELEMENT_ID] = std::to_string((OdUInt64)element->objectId().getHandle());;
```
[the remark this answers is not in the document]
Comment (sbanethia) [reply bouncer/src/repo/manipulator/modelconvertor/import/odaHelper/data_processor_rvt.cpp]: The changes is done.

**Comment by carmenfan** [comment]
let's park this. Closing the PR.

It's still very far away from being mergable.... for e.g, I'm not convinced we need to define our own OdInts for...

**Pull request closed without merging**